src/drivers/intel/cavs/ssp.c

```diff
@@ -376,71 +376,60 @@ static int ssp_set_config(struct dai *dai,
 		goto out;
 	}
 
-	/* MCLK config */
-	/* searching the smallest possible mclk source */
-	for (i = MAX_SSP_FREQ_INDEX; i >= 0; i--) {
+	clk_index = -1;
+
+	/* Searching for common smallest possible source for MCLK and BCLK */
+	for (i = 0; i < MAX_SSP_FREQ_INDEX; i++) {
+		/* MCLK config */
 		if (config->ssp.mclk_rate > ssp_freq[i].freq)
+			continue;
+
+		if (!(ssp_freq[i].freq % config->ssp.mclk_rate))
+			clk_index = i;
+		else
+			continue;
+
+		/* BCLK config */
+		if (ssp_freq[i].freq % config->ssp.bclk_rate) {
+			clk_index = -1;
+		} else {
+			mdivc |= MNDSS(ssp_freq[clk_index].enc);
+			mdiv = ssp_freq[clk_index].freq / config->ssp.bclk_rate;
+
+			/* select M/N output for bclk in case of Audio Cardinal
+			 * or PLL Fixed clock.
+			 */
+			if (ssp_freq[clk_index].enc !=
+				CLOCK_SSP_XTAL_OSCILLATOR)
+				sscr0 |= SSCR0_ECS;
 			break;
+		}
 
-		if (ssp_freq[i].freq % config->ssp.mclk_rate == 0)
+		/* check if we can get target BCLK with M/N */
```

Comment:
@lgirdwood Should we prefer higher clock without M/N or lower if possible with M/N? IDK if it's better to risk duty cycle or have higher clock.

Reply:
we should always favor the oscillator source
The PLL is not available in S0ix, and the M/N adds an unknown jitter
Also the oscillator clock is divided, and it's typically not useful to have an external MCLK togging at rates higher than ~20 MHz. So even if you take the 38.4 oscillator clock, then dividing it gives you a lower clock that the 24.576 MHz PLL for about the same results in terms of accuracy/power.

Put differently, the only reason to select the PLL is when the device requires a slot witdth that is a multiple of 16 or 32, which cannot be reached with the oscillator.

Reply:
@ranj063 @plbossart in that case code needs to be changed to have 2 loops instead of 1, as for now it will prefer m/n if it's possible with lower clock.

Reply:
@ranj063 you should reiterate with m/n possibilty only if no clock matched without m/n

Reply:
@jajanusz another possibility instead of adding 2 loops is to switch the order in which the clocks are added in ssp freq table. Today they are in ascending order but maybe they should be changed in the order XTAL, Cardinal and PLL?

BTW this change will actually fix the problem we're seeing with the TGL topology too without any changes to the SSP driver.